airflow/datasets/__init__.py

```diff
@@ -233,7 +265,15 @@ class _DatasetBooleanCondition(BaseDataset):
     def __init__(self, *objects: BaseDataset) -> None:
         if not all(isinstance(o, BaseDataset) for o in objects):
             raise TypeError("expect dataset expressions in condition")
-        self.objects = objects
+
+        expanded_objects = []
+        for obj in objects:
+            if isinstance(obj, Dataset):
+                expanded_objects.append(obj)
+            elif isinstance(obj, DatasetAlias):
+                expanded_objects.extend(obj.expand_datasets())
```

Comment:
I think this can just call `iter_datasets`?

Reply:
yep, makes sense. thanks!

Reply:
just tested, we'll need to do someting like `[ds for _, ds in obj.iter_datatset()]` so that's won't work

Reply:
The above looks fine to me though?

Reply:
but under the hook `iter_datatset` actually calls `expand_datasets`

in `iter_datatset`, we try to make `datasets` to `name, dataset` and then use `[ds for _, ds in obj.iter_datatset()]` to make it `datasets` again. looks like an extra step for me

Reply:
The plus side is you can use it for Dataset as well. Also this gets me wonder what happens if this gets fed one of the boolean conditionals?

Reply:
that's something I'm now trying to fix. I guess the original code is problematic. will fix it and address it all at once. thanks!